Modules/Devices/src/Devices.Infrastructure/Persistence/Repository/PnsRegistrationRepository.cs

```diff
@@ -20,8 +23,16 @@ public PnsRegistrationRepository(DevicesDbContext dbContext)
 
     public async Task Add(PnsRegistration registration, CancellationToken cancellationToken)
     {
-        await _registrations.AddAsync(registration, cancellationToken);
-        await _dbContext.SaveChangesAsync(cancellationToken);
+        try
+        {
+            await _registrations.AddAsync(registration, cancellationToken);
+            await _dbContext.SaveChangesAsync(cancellationToken);
+        }
+        catch (DbUpdateException exception)
+        {
+            if (exception.GetBaseException() is SqlException { Number: 2627 }) // violation in unique key index
```

Comment:
We have `DbUpdateExtensions` with a `HasReason` method in `Synchronization.Application`. You should move them to `BuildingBlocks.Application`, add a reason for this exception, update the implementation of `HasReason`, and use the extension method here.